### PR TITLE
Rspack

### PR DESCRIPTION
**Pull request by wendevlin**
## Proposed change

- change from webpack to rspack

## Type of change

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by MindFreeze** [comment]
I am getting these times for `build_frontend`:
```
Finished 'webpack-prod-app' after 3.35 min
Finished 'rspack-prod-app' after 1.3 min
```
And these for `develop`:
```
webpack Compiled successfully in 8.22s
rspack Compiled successfully in 5.60s
```
LGTM

**Review by MindFreeze — DISMISSED** [review]

File: build-scripts/rspack.cjs [hunk 1]
```diff
@@ -102,13 +99,18 @@ const createWebpackConfig = ({
...
-        // - Worklets use `importScripts`: https://github.com/webpack/webpack/issues/11543
-        chunks: (chunk) =>
-          !chunk.canBeInitial() &&
-          !new RegExp(`^.+-work${latestBuild ? "(?:let|er)" : "let"}$`).test(
-            chunk.name
-          ),
+        chunks: !isProdBuild
+          ? // improve incremental build speed, but blows up bundle size
+            new RegExp(
+              `^(?!.*(${Object.keys(entry).join("|")}|work(?:er|let))$)`
```
[the remark this answers is not in the document]
Comment (wendevlin) [reply build-scripts/rspack.cjs]: No change in bundle size :disappointed:

**Comment by MindFreeze** [comment]
The difference in bundle size with RegExp `chunks` is a bug in Rspack https://github.com/web-infra-dev/rspack/issues/8529

**Review by bramkragten — APPROVED** [review]

File: build-scripts/rspack.cjs [hunk 2]
```diff
@@ -168,10 +170,14 @@ const createWebpackConfig = ({
           stats: { assets: true, chunks: true, modules: true },
           transform: (stats) => JSON.stringify(filterStats(stats)),
         }),
-      !latestBuild &&
-        new TransformAsyncModulesPlugin({
```
[the remark this answers is not in the document]
Comment (steverep) [reply build-scripts/rspack.cjs:172]: AFAICT, `rspack` has the same problem that `webpack` does.  There are now a bunch of async functions in the legacy build that will cause errors on older browsers.

I'm not sure why this plugin was removed, but it should probably be added back.

[the remark this answers is not in the document]
Comment (steverep) [reply build-scripts/rspack.cjs:172]: ... or #23234 and then the transform isn't needed.

[the remark this answers is not in the document]
Comment (wendevlin) [reply build-scripts/rspack.cjs:172]: It wasn't supported by rspack, I removed it and tested it with browser stack if it is necessary. Maybe I did a mistake here, will check again or your solution solves it :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (steverep) [reply build-scripts/rspack.cjs:172]: No worries... you probably just didn't test anything that was old enough to not support async functions.  It's moot now that #23234 is merged.

I wrote that plugin so I'll have to see if I can quickly make it compatible with `rspack` for others.

[the remark this answers is not in the document]
Comment (wendevlin) [reply build-scripts/rspack.cjs:172]: Webpackbar did a rspack update, maybe you can take this as inspiration: 
- unjs/webpackbar#149

File: build-scripts/rspack.cjs [hunk 3]
```diff
@@ -102,7 +102,7 @@ const createRspackConfig = ({
         chunks: !isProdBuild
           ? // improve incremental build speed, but blows up bundle size
             new RegExp(
-              `^(?!.*(${Object.keys(entry).join("|")}|work(?:er|let))$)`
+              `^(?!(${Object.keys(entry).join("|")}|.*work(?:er|let))$)`
```
Comment (steverep) [inline build-scripts/rspack.cjs:105]: Why are we treating production and development builds differently?  Looks like the regex is trying to accomplish the same thing for the modern build, but will give a different result for the legacy build? 😕

Comment (wendevlin) [reply build-scripts/rspack.cjs:105]: We had troubles with the incremental build speed, so we found this solution but it fucks up all our chunks so for prod we still use the js lambda. Yeah you are right, maybe we should also disable this for legacyBuild

**Pull request merged**